File: src/kria/safety/rollback.py
"""
Rollback Manager
================
Creates timestamped snapshots of files (and registry keys) before any
RED-tier destructive action executes.  If the user says "undo the last
action", ``restore()`` reverses the change.

Storage layout:
  ~/.kria/rollback/
  └── 2026-04-11T14-30-00/
      ├── manifest.json        — what was changed + metadata
      └── files/               — file backups
          ├── original_name.txt
          └── ...

Retention: snapshots older than ``settings.rollback_retention_hours`` are
pruned automatically at startup and by the scheduled cleanup task.
"""
import hashlib
import json
import logging
import shutil
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Optional

from kria.infra.config import settings

logger = logging.getLogger("kria.safety.rollback")


class RollbackManager:
    def __init__(self) -> None:
        self._base = Path(settings.rollback_dir).expanduser()

    def _ensure_dir(self) -> None:
        self._base.mkdir(parents=True, exist_ok=True)

    # ── Snapshot creation ─────────────────────────────────────────

    async def create_snapshot(
        self,
        session_id: str,
        action: str,
        risk_level: str,
        files: list[str],
    ) -> Optional[str]:
        """
        Back up *files* into a new snapshot directory.
        Returns the snapshot ID (timestamp string) or None on failure.
        """
        try:
            self._ensure_dir()
            timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H-%M-%S")
            snap_dir = self._base / timestamp
            files_dir = snap_dir / "files"
            files_dir.mkdir(parents=True, exist_ok=True)

            changes = []
            for path_str in files:
                src = Path(path_str)
                if src.exists() and src.is_file():
                    # Avoid collisions with same filename from different dirs
                    unique_name = f"{src.parent.name}__{src.name}"
                    dest = files_dir / unique_name
                    shutil.copy2(src, dest)
                    file_hash = hashlib.sha256(src.read_bytes()).hexdigest()
                    changes.append({
                        "type": "file_backed_up",
                        "original_path": str(src),
                        "backup_path": str(dest),
                        "hash_sha256": file_hash,
                    })

            manifest = {
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "session_id": session_id,
                "action": action,
                "risk_level": risk_level,
                "changes": changes,
                "rollback_command": "restore_files",
                "expires": (
                    datetime.now(timezone.utc)
                    + timedelta(hours=settings.rollback_retention_hours)
                ).isoformat(),
            }
            (snap_dir / "manifest.json").write_text(
                json.dumps(manifest, indent=2), encoding="utf-8"
            )
            logger.info("Rollback snapshot created: %s (%d files)", timestamp, len(changes))
            return timestamp

        except Exception as exc:
            logger.error("Failed to create rollback snapshot: %s", exc)
            return None

    # ── Restore ───────────────────────────────────────────────────

    async def restore(self, rollback_id: str) -> bool:
        """Restore files from snapshot *rollback_id*. Returns True on success."""
        try:
            snap_dir = self._base / rollback_id
            manifest_path = snap_dir / "manifest.json"
            if not manifest_path.exists():
                logger.error("Rollback manifest not found: %s", rollback_id)
                return False

            manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
            restored = 0
            for change in manifest.get("changes", []):
                backup = Path(change["backup_path"])
                original = Path(change["original_path"])
                if backup.exists():
                    original.parent.mkdir(parents=True, exist_ok=True)
                    shutil.copy2(backup, original)
                    logger.info("Restored: %s", original)
                    restored += 1

            logger.info("Rollback complete: %d files restored from %s", restored, rollback_id)
            return True

        except Exception as exc:
            logger.error("Rollback restore failed for %s: %s", rollback_id, exc)
            return False

    # ── List snapshots ────────────────────────────────────────────

    def list_snapshots(self) -> list[dict]:
        if not self._base.exists():
            return []
        snapshots = []
        for entry in sorted(self._base.iterdir(), reverse=True):
            manifest_path = entry / "manifest.json"
            if manifest_path.exists():
                try:
                    m = json.loads(manifest_path.read_text(encoding="utf-8"))
                    snapshots.append({
                        "id": entry.name,
                        "action": m.get("action"),
                        "timestamp": m.get("timestamp"),
                        "expires": m.get("expires"),
                    })
                except Exception:
                    pass
        return snapshots

    # ── Cleanup ───────────────────────────────────────────────────

    async def cleanup_expired(self) -> int:
        """Remove snapshots past their expiry. Returns number removed."""
        if not self._base.exists():
            return 0
        now = datetime.now(timezone.utc)
        removed = 0
        for entry in list(self._base.iterdir()):
            if not entry.is_dir():
                continue
            manifest_path = entry / "manifest.json"
            try:
                if manifest_path.exists():
                    m = json.loads(manifest_path.read_text(encoding="utf-8"))
                    expires = datetime.fromisoformat(m["expires"])
                    if now > expires:
                        shutil.rmtree(entry)
                        logger.info("Pruned expired snapshot: %s", entry.name)
                        removed += 1
            except Exception as exc:
                logger.warning("Could not process snapshot %s: %s", entry.name, exc)
        return removed


rollback_manager = RollbackManager()

File: crates/kria-core/src/safety/rollback.rs
use std::path::{Path, PathBuf};
use chrono::Utc;
use sha2::{Sha256, Digest};

/// Manifest for a rollback snapshot.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct RollbackManifest {
    pub timestamp: String,
    pub session_id: String,
    pub action: String,
    pub risk_level: String,
    pub changes: Vec<ChangeRecord>,
    pub rollback_command: String,
    pub expires: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ChangeRecord {
    pub change_type: String,
    pub original_path: String,
    pub backup_path: String,
    pub hash_sha256: String,
}

/// Manages rollback snapshots for reversible safety.
///
/// Layout: {rollback_dir}/{timestamp}/manifest.json + files/
pub struct RollbackManager {
    rollback_dir: PathBuf,
    retention_hours: u64,
    max_storage_bytes: u64,
}

impl RollbackManager {
    pub fn new(rollback_dir: PathBuf, retention_hours: u64, max_storage_mb: u64) -> Self {
        let _ = std::fs::create_dir_all(&rollback_dir);
        Self {
            rollback_dir,
            retention_hours,
            max_storage_bytes: max_storage_mb * 1024 * 1024,
        }
    }

    /// Create a restore point before a destructive action.
    /// Returns the rollback ID (timestamp string).
    pub fn create_snapshot(
        &self,
        session_id: &str,
        action: &str,
        risk_level: &str,
        files_to_backup: &[&Path],
    ) -> anyhow::Result<String> {
        let ts = Utc::now().format("%Y-%m-%dT%H-%M-%S").to_string();
        let snapshot_dir = self.rollback_dir.join(&ts);
        let files_dir = snapshot_dir.join("files");
        std::fs::create_dir_all(&files_dir)?;

        let mut changes = Vec::new();

        for path in files_to_backup {
            if !path.exists() {
                continue;
            }
            let filename = path.file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .to_string();

            // Copy file to backup
            let backup_name = format!("{}_{}", changes.len(), &filename);
            let backup_path = files_dir.join(&backup_name);
            std::fs::copy(path, &backup_path)?;

            // Compute hash
            let data = std::fs::read(path)?;
            let mut hasher = Sha256::new();
            hasher.update(&data);
            let hash = format!("{:x}", hasher.finalize());

            changes.push(ChangeRecord {
                change_type: "file_backup".into(),
                original_path: path.to_string_lossy().into(),
                backup_path: backup_name,
                hash_sha256: hash,
            });
        }

        let expires = Utc::now() + chrono::Duration::hours(self.retention_hours as i64);
        let manifest = RollbackManifest {
            timestamp: ts.clone(),
            session_id: session_id.to_string(),
            action: action.to_string(),
            risk_level: risk_level.to_string(),
            changes,
            rollback_command: "restore_files".into(),
            expires: expires.to_rfc3339(),
        };

        let manifest_path = snapshot_dir.join("manifest.json");
        let json = serde_json::to_string_pretty(&manifest)?;
        std::fs::write(&manifest_path, json)?;

        tracing::info!(snapshot = %ts, action, "rollback snapshot created");
        Ok(ts)
    }

    /// Restore files from a snapshot.
    pub fn restore(&self, rollback_id: &str) -> anyhow::Result<Vec<String>> {
        let snapshot_dir = self.rollback_dir.join(rollback_id);
        let manifest_path = snapshot_dir.join("manifest.json");

        if !manifest_path.exists() {
            anyhow::bail!("rollback snapshot not found: {rollback_id}");
        }

        let json = std::fs::read_to_string(&manifest_path)?;
        let manifest: RollbackManifest = serde_json::from_str(&json)?;

        let mut restored = Vec::new();
        let files_dir = snapshot_dir.join("files");

        for change in &manifest.changes {
            let backup = files_dir.join(&change.backup_path);
            let original = Path::new(&change.original_path);

            if backup.exists() {
                // Ensure parent directory exists
                if let Some(parent) = original.parent() {
                    let _ = std::fs::create_dir_all(parent);
                }
                std::fs::copy(&backup, original)?;
                restored.push(change.original_path.clone());
            }
        }

        tracing::info!(snapshot = %rollback_id, count = restored.len(), "rollback restored");
        Ok(restored)
    }

    /// List all available snapshots.
    pub fn list_snapshots(&self) -> Vec<RollbackManifest> {
        let mut manifests: Vec<RollbackManifest> = Vec::new();
        if let Ok(entries) = std::fs::read_dir(&self.rollback_dir) {
            for entry in entries.flatten() {
                let manifest_path = entry.path().join("manifest.json");
                if manifest_path.exists() {
                    if let Ok(json) = std::fs::read_to_string(&manifest_path) {
                        if let Ok(m) = serde_json::from_str(&json) {
                            manifests.push(m);
                        }
                    }
                }
            }
        }
        manifests.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
        manifests
    }

    /// Prune expired snapshots and enforce storage limits.
    pub fn prune(&self) -> anyhow::Result<usize> {
        let now = Utc::now();
        let mut removed = 0;

        if let Ok(entries) = std::fs::read_dir(&self.rollback_dir) {
            let mut snapshots: Vec<_> = entries.flatten().collect();
            snapshots.sort_by(|a, b| a.file_name().cmp(&b.file_name()));

            for entry in &snapshots {
                let manifest_path = entry.path().join("manifest.json");
                if manifest_path.exists() {
                    if let Ok(json) = std::fs::read_to_string(&manifest_path) {
                        if let Ok(m) = serde_json::from_str::<RollbackManifest>(&json) {
                            if let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&m.expires) {
                                if expires < now {
                                    std::fs::remove_dir_all(entry.path())?;
                                    removed += 1;
                                }
                            }
                        }
                    }
                }
            }
        }

        // Enforce storage limit
        let total_size = self.total_size();
        if total_size > self.max_storage_bytes {
            // Remove oldest until under limit
            if let Ok(entries) = std::fs::read_dir(&self.rollback_dir) {
                let mut snapshots: Vec<_> = entries.flatten().collect();
                snapshots.sort_by(|a, b| a.file_name().cmp(&b.file_name()));
                for entry in snapshots {
                    if self.total_size() <= self.max_storage_bytes {
                        break;
                    }
                    std::fs::remove_dir_all(entry.path())?;
                    removed += 1;
                }
            }
        }

        Ok(removed)
    }

    fn total_size(&self) -> u64 {
        walkdir::WalkDir::new(&self.rollback_dir)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter_map(|e| e.metadata().ok())
            .map(|m| m.len())
            .sum()
    }
}
